heron/uploaders/tests/java/com/twitter/heron/uploader/gcs/GcsUploaderTests.java

```diff
@@ -0,0 +1,142 @@
+        .createObject(Mockito.matches(topologyObjectName), Mockito.any(File.class)))
+        .thenReturn(createObject(topologyObjectName));
+
+    uploader.initialize(createDefaultBuilder().build());
+
+    String expectedUri =
+        String.format("https://storage.googleapis.com/%s/%s/%s",
+            bucket, topologyName, topologyPackageName);
+    assertEquals(new URI(expectedUri), uploader.uploadPackage());
+  }
+
+  @Test
+  public void verifyObjectBackedUpIfExists() throws IOException {
+    // return an object to simulate that the topology has been uploaded before
+    final StorageObject currentObject = createObject(topologyObjectName);
+    Mockito.when(mockGcsController
+        .getObject(Mockito.matches(topologyObjectName)))
+        .thenReturn(currentObject);
+
+    // return an object when we try to create one
+    Mockito.when(mockGcsController
+        .createObject(Mockito.matches(topologyObjectName), Mockito.any(File.class)))
+        .thenReturn(createObject(topologyObjectName));
+
+    uploader.initialize(createDefaultBuilder().build());
+
+    uploader.uploadPackage();
+
+    // verify that we copied the old topology before uploading the new one
+    Mockito.verify(mockGcsController)
+        .copyObject(topologyObjectName, previousTopologyObjectName, currentObject);
+  }
+
+  @Test
+  public void restorePreviousVersionOnUndo() throws IOException {
+    final StorageObject previousObject = createObject(previousTopologyObjectName);
+    Mockito.when(mockGcsController
+        .getObject(Mockito.matches(previousTopologyObjectName)))
+        .thenReturn(previousObject);
+
+    uploader.initialize(createDefaultBuilder().build());
+
+    uploader.undo();
+
+    // verify that we restored the previous topology
+    Mockito.verify(mockGcsController)
+        .copyObject(previousTopologyObjectName, topologyObjectName, previousObject);
+  }
+
+  @Test
+  public void doNotRestorePreviousVersionIfItDoesNotExist() throws IOException {
+    Mockito.when(mockGcsController
+        .getObject(Mockito.matches(previousTopologyObjectName)))
+        .thenReturn(null);
+
+    uploader.initialize(createDefaultBuilder().build());
+
+    uploader.undo();
+
+    Mockito.verify(mockGcsController, Mockito.never())
+        .copyObject(Mockito.anyString(), Mockito.anyString(), Mockito.any(StorageObject.class));
+  }
+
+  private StorageObject createObject(String name) {
+    final StorageObject storageObject = new StorageObject();
+    storageObject.setName(name);
+    return storageObject;
+  }
+
+  private Config.Builder createDefaultBuilder() {
+    return Config.newBuilder()
+        .put(GcsContext.HERON_UPLOADER_GCS_BUCKET, bucket)
+        .put(Key.TOPOLOGY_NAME, topologyName)
+        .put(Key.TOPOLOGY_PACKAGE_FILE, topologyPackageName);
+  }
+
+  private GcsUploader createUploaderWithMockController() {
+    return new GcsUploader() {
+      @Override
+      GcsController createGcsController(Config configuration, String storageBucket) {
```

Comment:
instead of subclassing, could you instead return a mock GcsUploader that has the createGcsController method mocked to return the mock GcsController. better to consistently use mocks than a mix of mocks and sub-classes.